strapi/strapi
# Let custom fields specify a custom input size

### What does it do?

Extends the custom field server registry API to add a new optional inputSize property using this format:

```js
strapi.customFields.register({
    name: 'color',
    plugin: 'color-picker',
    type: 'string',
    // new (and optional)
    inputSize: {
      default: 4,
      isResizable: true,
    },
  });
```

It applies some validation: if `inputSize` is provided, then it must be an object that must have both a `default` size and the `isResizable` boolean. The `default` is a number referring to the amount of columns that the field will occupy in the 12-col grid. It can either be 4, 6, 8 or 12.

### Why is it needed?

Because currently custom fields obey by the rules of their underlying type. If a custom field uses JSON, it's always 12 columns and non resizable. But this doesn't always make sense, as the UI for a custom field is often very different from the UI of the Strapi input for the underlying type.

For example, a custom field could visually only show a simple select, but save its data as JSON. In that case forcing it to look full-width is not a good UX.

### How to test it?

Try creating a custom field and giving it a custom `inputSize` object. Or just modify the existing color picker one, like in the snippet above. Then build your app before running `yarn develop`, or use `--watch-admin`, since port 4000 doesn't handle custom fields well.

- when you add an attribute using this custom field, and open the content manager, the default size you entered should be applied
- when you go to "configure the view", and try to edit that field, your `isResizable` value should be applied
- in "configure the view", delete a custom field input, then recreate it. The default size should you the one you specified
- try modifying your custom field `inputSize` to give it invalid values, it should throw errors
- do some sanity checks on non-custom field attributes (create them and check their default size, try to resize them etc.). They should behave as they did before

### Related issue(s)/PR(s)

This PR was made possible by #16387 which shipped in 4.9.2
